src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+{
+    int axis = -1;
+    for(auto&& input : ins->inputs())
+    {
+        if(is_common.find(input) != is_common.end())
+        {
+            int cur_axis = find_unique_axis(ins, input);
+            if((cur_axis == -1) || ((axis != -1) && (cur_axis != axis)))
+                return -1;
+            axis = cur_axis;
+        }
+    }
+    if(is_conv(ins) && (axis != get_channel_axis()))
+        return -1;
+
+    return axis;
+}
+
+// remove instructions in given vector except the first one.
+void horizontal_fusion_impl::remove_redundant_roots(std::vector<instruction_ref>& base_instrs)
+{
+    instruction_ref root_ins = base_instrs.at(0);
+    for(unsigned long ndx = 1; ndx < base_instrs.size(); ndx++)
+    {
+        instruction_ref base                 = base_instrs.at(ndx);
+        std::vector<instruction_ref> outputs = base->outputs();
+        for(auto&& output : outputs)
+            instruction::replace_argument(output, base, root_ins, false);
+        p_program->remove_instruction(base);
+    }
+}
+
+// update hash tree.
+void horizontal_fusion_impl::update_hash_tree(unsigned hash_id)
+{
+    unsigned first = *(hash_instrs[hash_id].begin());
+    hash_instrs[hash_id].clear();
+    hash_instrs[hash_id].insert(first);
+}
+
+instruction_ref horizontal_fusion_impl::break_split(int enum_ndx, instruction_ref split_ins)
+{
+    const operation& op = split_ins->get_operator();
+    int first           = (any_cast<op::split>(op)).slice_selector.first;
+    int second          = (any_cast<op::split>(op)).slice_selector.second;
+    if((first < 0) || (second < first))
+        MIGRAPHX_THROW("unexpected selector");
+    if((enum_ndx != first) && (enum_ndx != second))
+        MIGRAPHX_THROW("unexpected slice enumerator");
+
+    if(first == second)
+        return split_ins;
+    int axis                    = (any_cast<op::split>(op)).axis;
+    std::vector<int> slice_dims = (any_cast<op::split>(op)).slice_dims;
+    instruction_ref input       = split_ins->inputs().at(0);
+    instruction_ref new_split   = p_program->insert_instruction(
+        split_ins, op::split{axis, slice_dims, {enum_ndx, enum_ndx}}, input);
+    operation& op_edit = split_ins->get_operator_edit();
+
+    if(first == enum_ndx)
+        (any_cast<op::split>(op_edit)).slice_selector.first = enum_ndx + 1;
+    else
+        (any_cast<op::split>(op_edit)).slice_selector.second = enum_ndx - 1;
+
+    std::vector<shape> shapes;
+    shapes.push_back(input->get_shape());
+    shape new_shape = (any_cast<op::split>(op_edit)).compute_shape(shapes);
+    split_ins->set_shape(new_shape);
+    return new_split;
+}
+
+bool horizontal_fusion_impl::has_unique_output(const std::vector<instruction_ref>& instrs)
+{
+    std::unordered_map<instruction_ref, bool> seen;
+    for(auto&& ins : instrs)
+    {
+        for(auto&& output : ins->outputs())
+        {
+            if(seen.find(output) != seen.end())
+                return false;
+            seen[output] = true;
```

Comment:
I think this can be written as:

```cpp
            if(not seen.emplace(output, true).second)
                return false;
```

And it will avoid a second search.

Reply:
fixed.